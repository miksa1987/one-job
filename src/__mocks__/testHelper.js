export const saveTodoMock = jest.fn();