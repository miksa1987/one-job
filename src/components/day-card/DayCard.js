import React from 'react';
import styled from 'styled-components';
import SetTodoTime from './SetTodoTime';
import TodoStore from '../../store';
import Firebase from '../../firebase/firebase';
import useField from '../../hooks/useField';
import moment from 'moment';
import { observer } from 'mobx-react-lite';
import { H2 } from '../common/Headers';

const CardBase = styled.div`
  display: grid;
  grid-template: 3rem 6rem 6rem 4rem / 80vh;
  justify-content: center;
  background-color: #eceff4;
  border-bottom: 1px solid #c8cee9;
  padding: 10px;
`;

const DayCard = observer(() => {
  const [ todoText, setTodoText ] = useField('text');
  const [ reflectText, setReflectText ] = useField('text');

  const store = React.useContext(TodoStore);
  const firebase = React.useContext(Firebase);
  const currentTodo = store.todo;

  React.useEffect(() => {
    if (currentTodo) {
      setTodoText(currentTodo.task);
      setReflectText(currentTodo.reflect);
    }
  }, [currentTodo]);

  const saveTodo = async () => {
    const newTodo = {
      task: todoText.value,
      reflect: reflectText.value,
      date: moment(store.date, 'YYYY-MM-DD').format('YYYY-MM-DD')
    }
    const key = currentTodo.key || null
    await firebase.saveTodo(newTodo, key);
  }

  return (
    <CardBase>
      <H2>Today<button onClick={saveTodo}>save [TEMPORARY SOLUTION]</button></H2>
      <textarea {...todoText} />
      <textarea {...reflectText} />      
      <SetTodoTime />
    </CardBase>
  );
});

export default DayCard;