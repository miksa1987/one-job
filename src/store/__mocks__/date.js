export default class DateHandlerMock {
  setDate = jest.fn();
  setNextDayDate = jest.fn();
  setPreviousDayDate = jest.fn();
  setCurrentDayDate = jest.fn();
}