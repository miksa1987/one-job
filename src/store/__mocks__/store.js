import { decorate, computed, observable } from 'mobx';
import { createContext } from 'react';
import { 
  saveTodoMock, 
  setCurrentDateMock,
  loginAndSetUserMock,
  createAndSetUserMock,
  setNotificationMock 
} from './testHelper';

class TodoStore {
  currentTodo = {
    task: 'test',
    reflect: 'test',
    date: '2019-11-13',
    time: [ 12, 0 ],
    key: '1'
  };
  currentDate = '2019-11-13';

  setCurrentDate = (date) => {
    setCurrentDateMock();
    this.currentDate = date;
  }

  saveTodo = (newTodo, key) => {
    saveTodoMock();
    this.currentTodo = { ...newTodo, key: '1' };
  }

  setNextDayDate = () => this.currentDate = '2019-11-14';
  setPreviousDayDate = () => this.currentDate = '2019-11-12';
  setCurrentDayDate = () => this.currentDate = '2019-11-13';
  
  seeIfTodoTimeIsPassed = () => {}
  setNotification = () => setNotificationMock();

  loginAndSetUser = () => loginAndSetUserMock();
  createAndSetUser = () => createAndSetUserMock();

  get user() {
    return {};
  }
  
  get todo() {
    return this.currentTodo;
  }

  get date() {
    return this.currentDate;
  }

  get allTodos() {
    return [ this.currentTodo, this.currentTodo, this.currentTodos ];
  }

  get todoTimePassed() {
    return true;
  }
}

decorate(TodoStore, {
  currentUser: observable,
  currentTodo: observable,
  currentDate: observable,
  todos: observable,
  user: computed,
  todo: computed,
  date: computed,
  allTodos: computed
});

export default createContext(new TodoStore());