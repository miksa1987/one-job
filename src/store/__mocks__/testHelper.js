export const saveTodoMock = jest.fn();
export const setCurrentDateMock = jest.fn();
export const createAndSetUserMock = jest.fn();
export const loginAndSetUserMock = jest.fn();
export const setNotificationMock = jest.fn();