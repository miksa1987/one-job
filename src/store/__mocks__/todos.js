export default class TodosHandlerMock {
  getAndSetTodos = jest.fn();
  setCurrentTodo = jest.fn();
  checkForTodo = jest.fn();
  saveTodo = jest.fn();
  seeIfTodoTimeIsPassed = jest.fn();
}