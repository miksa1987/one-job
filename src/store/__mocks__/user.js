export default class UserHandlerMock {
  createAndSetUser = jest.fn();
  loginAndSetUser = jest.fn();
  logoutUser = jest.fn();
  setUser = jest.fn();
}