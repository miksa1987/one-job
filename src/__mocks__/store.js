import { decorate, computed, observable } from 'mobx';
import { createContext } from 'react';
import { saveTodoMock } from './testHelper';

class TodoStore {
  currentTodo = {
    task: 'test',
    reflect: 'test',
    date: '2019-11-13',
    time: [ 12, 0 ],
    key: '1'
  };

  saveTodo = (newTodo, key) => {
    saveTodoMock();
    this.currentTodo = { ...newTodo, key };
  }

  get user() {
    return {};
  }
  
  get todo() {
    return this.currentTodo;
  }

  get date() {
    return '2019-11-13';
  }

  get allTodos() {
    return [ this.currentTodo, this.currentTodo, this.currentTodos ];
  }
}

decorate(TodoStore, {
  currentUser: observable,
  currentTodo: observable,
  currentDate: observable,
  todos: observable,
  user: computed,
  todo: computed,
  date: computed,
  allTodos: computed
});

export default createContext(new TodoStore());