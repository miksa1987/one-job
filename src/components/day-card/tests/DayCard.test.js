import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DayCard from '../DayCard';

import { saveTodoMock } from '../../../__mocks__/testHelper';
jest.mock('../../../store');

describe('DayCard', () => {
  afterEach(cleanup);
  
  test('Everything gets rendered', () => {
    const component = render(<DayCard />);
    const oneJobText = component.container.querySelector('#one-job-text');
    const todoText = component.container.querySelector('#todo-text');
    const reflectText = component.container.querySelector('#reflect-text');

    expect(oneJobText).toHaveTextContent('What is your one job today?');
    expect(todoText).toHaveTextContent('test');
    expect(reflectText).toHaveTextContent('test');
  });

  test('Todo text gets saved(saveTodo is called)', () => {
    saveTodoMock.mockClear();
    const component = render(<DayCard />);
    
    const todoText = component.container.querySelector('#todo-text');
    const reflectText = component.container.querySelector('#reflect-text');

    fireEvent.click(todoText);
    fireEvent.change(todoText, { target: { value: 'changed' } });
    fireEvent.focus(reflectText);
    fireEvent.blur(todoText);

    expect(saveTodoMock.mock.calls.length).toBe(1);
  });
  
  test('Reflect text gets saved(saveTodo is called)', () => {
    saveTodoMock.mockClear();
    const component = render(<DayCard />);
    
    const todoText = component.container.querySelector('#todo-text');
    const reflectText = component.container.querySelector('#reflect-text');

    fireEvent.click(reflectText);
    fireEvent.change(reflectText, { target: { value: 'changed' } });
    fireEvent.focus(todoText);
    fireEvent.blur(reflectText);

    expect(saveTodoMock.mock.calls.length).toBe(1);
  });

  test('Set reminder time saves the todo', () => {
    saveTodoMock.mockClear();
    const component = render(<DayCard />);
    
    const timeHours = component.container.querySelector('#todo-time-hours');
    const setTime = component.container.querySelector('#todo-time-set');

    fireEvent.change(timeHours, { target: { value: 23 } });
    fireEvent.click(setTime);

    expect(saveTodoMock.mock.calls.length).toBe(1);
  });
});