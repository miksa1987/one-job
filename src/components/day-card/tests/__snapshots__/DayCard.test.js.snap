// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayCard Everything gets rendered 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-ifAKCX coOVlB"
      >
        <h2
          id="one-job-text"
        >
          What is your one job today?
        </h2>
        <textarea
          id="todo-text"
          type="text"
        >
          test
        </textarea>
        <strong>
          Now, reflect on how you felt doing the job:
        </strong>
        <textarea
          id="reflect-text"
          type="text"
        >
          test
        </textarea>
        <div
          class="sc-bdVaJa jpqvBK"
        >
          <div
            class="sc-bxivhb izQvlR"
          >
            <strong>
              When will you do it? 
            </strong>
            <input
              class="sc-htpNat hQIzVp"
              id="todo-time-hours"
              type="number"
              value="12"
            />
            <input
              class="sc-htpNat hQIzVp"
              id="todo-time-minutes"
              type="number"
              value="0"
            />
            <button
              class="sc-bwzfXH hEtCYp"
              id="todo-time-set"
            >
              Set
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-ifAKCX coOVlB"
    >
      <h2
        id="one-job-text"
      >
        What is your one job today?
      </h2>
      <textarea
        id="todo-text"
        type="text"
      >
        test
      </textarea>
      <strong>
        Now, reflect on how you felt doing the job:
      </strong>
      <textarea
        id="reflect-text"
        type="text"
      >
        test
      </textarea>
      <div
        class="sc-bdVaJa jpqvBK"
      >
        <div
          class="sc-bxivhb izQvlR"
        >
          <strong>
            When will you do it? 
          </strong>
          <input
            class="sc-htpNat hQIzVp"
            id="todo-time-hours"
            type="number"
            value="12"
          />
          <input
            class="sc-htpNat hQIzVp"
            id="todo-time-minutes"
            type="number"
            value="0"
          />
          <button
            class="sc-bwzfXH hEtCYp"
            id="todo-time-set"
          >
            Set
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
